File: solutions/javascript/clock/1/clock.js

function pad(n, size) {
  let result = n.toFixed(0);
  if (result.length < size) {
    result = "0000000000".substring(0, size-result.length) + result;
  }
  return result;
}

export class Clock {

  static DAY_IN_MINUTES = 24 * 60;

  constructor(hours=0, minutes=0) {
    
    let timeInMinutes = (hours * 60) + minutes;
    // normalize the time between 0 hours and 24 hours
    timeInMinutes = timeInMinutes % Clock.DAY_IN_MINUTES;   
    if (timeInMinutes < 0) timeInMinutes += Clock.DAY_IN_MINUTES;
    this._timeInMinutes = timeInMinutes;
  }

  toString() {
    let hours = Math.trunc(this._timeInMinutes / 60);
    let minutes = this._timeInMinutes - hours * 60;
    return `${pad(hours, 2)}:${pad(minutes, 2)}`;
  }

  plus(minutes) {
    return new Clock(0, this._timeInMinutes + minutes);
  }

  minus(minutes) {
    return new Clock(0, this._timeInMinutes - minutes);
  }

  equals(other) {
    return this._timeInMinutes === other._timeInMinutes;
  }
}

File: solutions/lua/clock/1/clock.lua
local Clock = {}
Clock.__index = Clock

function Clock.at(hours, minutes)

    local minutes = minutes or 0

    local time = {}
    setmetatable(time, Clock)
    local h = minutes // 60
    time.minutes = minutes - h * 60
    time.hours = (hours + h) % 24

    return time
end

function Clock:plus(minutes)

    return Clock.at(self.hours, self.minutes + minutes)

end

function Clock:equals(other)

    return self.hours == other.hours and self.minutes == other.minutes
end

function Clock:minus(minutes)

    -- the constructor 'at' can deal with negative hours but not negative minutes
    -- if the number of minutes is negative, switch to the previous hour
    local remaining_minutes = self.minutes - minutes
    local hours = self.hours
    while remaining_minutes < 0 do
        remaining_minutes = remaining_minutes + 60
        hours = hours - 1
    end

    return Clock.at(hours, remaining_minutes)

end

function Clock:__tostring()
    return string.format('%02d:%02d', self.hours, self.minutes)
end

return Clock
